modules/backend/formwidgets/FileUpload.php

```diff
@@ -337,6 +343,9 @@ public function onRemoveAttachment()
         $fileModel = $this->getRelationModel();
         if (($fileId = post('file_id')) && ($file = $fileModel::find($fileId))) {
             $this->getRelationObject()->remove($file, $this->sessionKey);
+            if ($this->deleteOnDetach) {
+                $file->delete();
+            }
```

Comment:
@goldmont As mentioned by @bennothommo this probably has to be implemented in the base File model itself to be run after detaching is committed by the deferred binding process.